src/microbe_stage/editor/MicrobeEditorAction.cs

```diff
@@ -43,13 +43,22 @@ public class MicrobeEditorAction : ReversibleAction
 
     public override void DoAction()
     {
-        editor.ChangeMutationPoints(-Cost);
-        redo(this);
+        redo(this, editor.CurrentSpecies);
+        editor.UpdateMutationPoints();
+        editor.UpdateGUI();
     }
 
     public override void UndoAction()
     {
-        editor.ChangeMutationPoints(Cost);
-        undo(this);
+        undo(this, editor.CurrentSpecies);
+        editor.UpdateMutationPoints();
+        editor.UpdateGUI();
+    }
+
+    public int Forcast()
+    {
+        var prospectiveSpecies = (MicrobeSpecies)editor.CurrentSpecies.Clone();
+        redo(this, prospectiveSpecies);
+        return editor.MutationPointsAfterChange(prospectiveSpecies);
```

Comment:
I think I would *vastly* prefer an explicit cost calculation callback instead of doing this hack with an unnecessary (expensive) clone operation.

Reply:
This isn't a hack, this was the whole point of wanting to do this change. What is the mutation cost of going from species A to species B? The result of the MutationPointsAfterChange function. One definition, no duplicate logic to make bugs like what the editor has currently. If a clone operation is expensive, that's a much deeper problem (and one I haven't observed in practice).

Reply:
It isn't the most expensive operation currently. However, we are making a game where you can't slip up performance too much before it impacts the performance.
A much better option would be to apply the operation on the species, measure the effect, and then rollback the operation (there must be an always working undo anyway) to skip on allocating unnecessary objects.

Reply:
That sounds like a strategy that might actually lead to performance issues, since that will take O(n) time with respect to the size of the action stack in the editor, which can grow arbitrarily large, whereas cloning should only become more costly with larger microbes, and with a non-graphical model object should never be more than a couple milliseconds. That also breaks referential transparency, and makes the function not thread-safe, which may be a concern when called by UI.

Another note is that to implement the cost algorithm, the code would need to make a copy of the old organelle positions for comparisons anyway. Since that's most of what a MicrobeSpecies is, that's going to mitigate most of these performance bonuses.

Reply:
But you need to at most apply and revert the new action? You don't need to rewind the entire action stack ever, unless the player hammers the undo and redo keys.

Any action needs to be reversible fully so, you can always do the following:
- Make one clone for use as the edited species in the editor
- For calculating the cost / attempting an action, first run the action on the edited species, check the cost against the non-modified original species instance, then if needed reverse the applied action on the edited species to get back to the edited species state before applying the action.

Is that not correct?

Reply:
You would need to run the cost function twice, since the cost of a marginal action is the difference between the cost of the new edited cell minus the cost of the current edited cell. These are redundant calculations with the last and next time a change is made, so by storing these somewhere in the editor you could cut those down, but now you've got contained objects directly modifying the parent, which is a big violation of dependency inversion. None of these strategies are thread-safe; if you ever wanted to do something like having buttons gray out if you can't afford the action, you will need to queue calls to check the cost or risk permanently changing the current microbe.

The alternative is to make a fresh instance that only this object needs to worry about, using a clone method that auto-evo runs 10 times per species per patch already without performance issues (we've seen plenty of speed-runners doing double-evolutions, so it's not just because it's on a background thread). I think that modifying a shared instance would be introducing real risks in order to avoid an imagined performance problem. 

Sorry for being abstinent, but I believe this is an important design choice that will come back to bite you.

Reply:
I'm hesitant to suggest this due to the added complexity and potential for errors, but it seems like the only proper thing would be to write cost computation functions for each of the actions separately, basically making a variant of each of the actions that just computes the changes it would make and returns the MP cost. But as I said that is probably too error prone to be worth it.

So in the end I guess as much as it hurts my C++ roots (especially as the clone operation gets more expensive the more organelles a species contains), the clone action is the best design we have come up for here.

Just one more note about the auto-evo, there it really does not matter how expensive the cloning operation is, as it doesn't need to be able to run within a single frame to tell the player how much removing an organelle would cost, whereas this definitely needs to be able to do that.

I'll be fine with the performance if you test that a species with 3000 organelles doesn't cause performance issues.

Reply:
Alright, you got me there. With this change you start to notice lag at around the 1500 mark. If that's really your requirements (and yes I realize some people will have slower computers) then this won't work.

Reply:
Well the new membrane code will have a limit of 3000 organelles. So other editor code also should be coded to work up to that limit. (It may actually be hexes, but in practice I think other code should be tested with 3000 organelles as there are many single hex organelles, so that's the worst case scenario).

Reply:
One thing to check before reworking stuff is to test on master branch how much lag you get there. There is (especially with 6 way symmetry) slight performance degradation when placing organelles near the 3000 amount and above.

Reply:
I gave that a try (and remembered the symmetry system exists the second time which saved me a lot of time), and it ran smoothly up to 2000 on my machine. With this branch you can still reach 3000 without crashing, but you'll have to wait between placements for the lag to catch up.

Reply:
(You can make a save to easily test the different versions, or at least hopefully if this PR doesn't break save compatibility. If this does break compatibility I've been thinking of adding an automatic save upgrade system for a while...).
It would be nice to be able to place the 3000th organelle without more than a small lag spike (maybe like up to 200ms or so which doesn't feel too terrible yet).

Reply:
Bad news on both fronts. This PR does break save compatibility (I assume because the editor has different variables), and at the 3k mark people with lower-end boxes should be expecting 1500ms lag per organelle added. If you really feel that the code needs to be handling microbes thousands of organelles in size, then your initial performance concerns are legitimate and this algorithm won't work. Frankly, if that's what the editor needs to do, I would rather focus my time on other areas of the code.

Reply:
It would be nice if the game wouldn't choke on 3000 organelles...
Maybe someone else has opinions on this?
I've seen many screenshots where many players enjoy making ridiculously large cells, so I really wouldn't want to take that away from them...

Here's the issue about the save upgrade system: https://github.com/Revolutionary-Games/Thrive/issues/2152

Reply:
the 3000 hex limit is almost completely arbitrary. If need be I can always change it to something lower and regenerate my lookup tables, it will not really effect my geometry calculations. a 2000hex cell is still quite massive.

Reply:
You know what I just realized? It's stupid easy to just not run this code if the player is in freebuild. Now this will only be a problem for players who want to play a normal game for 666 generations, each time spending all their mutation points to buy 4 cytoplasm.

Reply:
Yeah, that is certainly an ingenious solution. 
I can't think of any situation (even in ascension stage) where we'd need to calculate the MP cost when the player can in a reasonable amount of time place so many organelles to cause problems.

Reply:
he...he's got a point